### PR TITLE
Test: Refactor tests and add pmem allocator test cases

### PR DESCRIPTION
**Pull request by sherry-1001**
### What is changed and how it works?
     (1) Refactor tests
     (2) Add PMem Allocator test cases: Different configs; Fragmentation; Free list.
     (3) Fix some bugs
          Now we have a bug: When thread num is 1, we call `pmem_allocator->Free` function, then call `pmem_allocator->Background`,  the memory fragmentation have not be merged. Please see `TestPMemFragmentation`. However I used `TestPMemAlocFreeList`,  see the memory fragmentation can be merged.  @JiayuZzz 

- Unit test
    Add pmem allocator test.
    Refactor tests: integration same codes
- Integration test
- No code

Side effects
    No

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# For kvdk api general test
+set(TEST_SOURCES tests.cpp)
+add_executable(dbtest ${TEST_SOURCES})
+target_link_libraries(dbtest PUBLIC engine gtest gmock gtest_main)
```
Comment (ZiyanShi) [inline tests/CMakeLists.txt:4]: Do we need gmock?

File: tests/test_pmem_allocator.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  uint64_t alloc_size = 8;
+
+  // params config
+  std::vector<uint64_t> num_segment_blocks{1024, 2 * 1024, 2 * 1024 * 1024};
+  std::vector<uint32_t> block_sizes{16, 32, 64};
+  std::vector<uint32_t> num_threads = {1, 16};
+
+  for (int i = 0; i < num_segment_blocks.size(); ++i) {
+    for (auto num_thread : num_threads) {
+      thread_manager_.reset(new (std::nothrow) ThreadManager(num_thread));
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:60]: Check nullptr for nothrow, or just let allocator do its job -- throw std::bad_alloc

File: tests/test_pmem_allocator.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+          records.push_back(space_entry);
+        }
+
+        auto alloc_bytes = pmem_alloc->PMemUsageInBytes();
+
+        for (uint64_t j = 0; j < records.size(); ++j) {
+          pmem_alloc->Free(records[j]);
+        }
+
+        ASSERT_EQ(pmem_alloc->PMemUsageInBytes(), (std::int64_t)0);
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:83]: 0LL

File: tests/test_pmem_allocator.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  std::deque<SpaceEntry> records;
+  thread_manager_.reset(new ThreadManager(num_thread));
+  PMEMAllocator *pmem_alloc = PMEMAllocator::NewPMEMAllocator(
+      pmem_path, pmem_size, num_segment_block, block_size, num_thread, true,
+      false, nullptr);
+  ASSERT_NE(pmem_alloc, nullptr);
+
+  thread_manager_->MaybeInitThread(access_thread);
+  // allocate 1024 bytes
+  records.push_back(pmem_alloc->Allocate(alloc_size[0]));
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:168]: Use the number may be more clear?

File: tests/tests.cpp [hunk 5]
```diff
@@ -61,6 +75,187 @@ class EngineBasicTest : public testing::Test {
...
+    key = "", val = "val";
+    ASSERT_EQ(SetFunc(collection, key, val), Status::Ok);
+    ASSERT_EQ(GetFunc(collection, key, &got_val), Status::Ok);
+    ASSERT_EQ(val, got_val);
+    ASSERT_EQ(DeleteFunc(collection, key), Status::Ok);
+    ASSERT_EQ(GetFunc(collection, key, &got_val), Status::NotFound);
+    engine->ReleaseAccessThread();
+  }
+
+  std::function<void(uint32_t)> BasicOperator(const std::string &collection,
```
Comment (ZiyanShi) [inline tests/tests.cpp:218]: CreateBasicOperationTest

**Review by ZiyanShi — APPROVED** [review]

File: tests/test_pmem_allocator.cpp (hunk 2 above)
Comment (sherry-1001) [reply tests/test_pmem_allocator.cpp:60]: remove std::nothrow

File: tests/tests.cpp (hunk 5 above)
Comment (sherry-1001) [reply tests/tests.cpp:218]: ok

File: tests/test_pmem_allocator.cpp (hunk 4 above)
Comment (sherry-1001) [reply tests/test_pmem_allocator.cpp:167]: ok

File: tests/CMakeLists.txt (hunk 1 above)
Comment (sherry-1001) [reply tests/CMakeLists.txt]: removed

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+message(${PROJECT_SOURCE_DIR})
+# For pmem allocator module tests
+set(TEST_PMEM_ALLOC test_pmem_allocator.cpp)
+add_executable(dbtest_pmem_allocator ${TEST_PMEM_ALLOC})
+target_include_directories(dbtest_pmem_allocator 
+    PRIVATE 
+    ${PROJECT_SOURCE_DIR}/engine 
+    ${PROJECT_SOURCE_DIR}/include
+    )
+target_link_libraries(dbtest_pmem_allocator PUBLIC engine gtest gmock gtest_main)
```
Comment (ZiyanShi) [inline tests/CMakeLists.txt:20]: remove gmock?

File: tests/tests.cpp [hunk 7]
```diff
@@ -61,6 +75,247 @@ class EngineBasicTest : public testing::Test {
...
+          next = k;
+        }
+      }
+      ASSERT_EQ(entries, 0);
+      engine->ReleaseSortedIterator(iter);
+    };
+    LaunchNThreads(configs.max_access_threads, IteratingThrough);
+  }
+
+  void UnorderedSeekIterator(const std::string &collection,
```
Comment (ZiyanShi) [inline tests/tests.cpp:198]: You cannot Seek in hashes.

Comment (ZiyanShi) [reply tests/tests.cpp:198]: TestSortedIterator
TestHashesIterator

**Review by JiayuZzz — APPROVED** [review]

File: tests/tests.cpp [hunk 8]
```diff
@@ -61,6 +75,247 @@ class EngineBasicTest : public testing::Test {
...
+      // Update
+      AssignData(val2, fast_random_64() % 1024);
+      ASSERT_EQ(SetFunc(collection, key2, val2), Status::Ok);
+      ASSERT_EQ(GetFunc(collection, key2, &got_val2), Status::Ok);
+      ASSERT_EQ(got_val2, val2);
+    }
+  }
+
+  void ForwardIter(Iterator *iter, Types type) {
+    int forward_entries = 0;
```
Comment (ZiyanShi) [inline tests/tests.cpp:293]: Unused?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

**Review by ZiyanShi — APPROVED** [review]

**Pull request merged**